### ISSUE TITLE
Default options for filter types

### ISSUE THREAD

**Issue by RomainSanchez** [sonata-project/SonataAdminBundle#4255]
Hi, here is my use case: 

I need a custom query for a `doctrine_orm_admin_filter` and it works fine by passing a callback field option.

However, i'd like to be able to make a type that would have this callback by default as it is generic so that i can use it as filter type directly without providing the same callback in `field_options` every where i need it (various admins).

I tried making a form type extending sonata_type_model with the callback defined as default to use it in the filter.  
But the problem is 
    `HelperController::retrieveAutocompleteItemsAction()` 
tries to retieve the callback from the filter description.So i tried defining the callback in the filter type (inherited from ModelAutocompleteFilter) in 
   ` getDefaultOptions()` ,
    `initialize()` 
and even in the constructor but to no avail.

Bottom line: would it be possible to have defaults options in filter types as in form types (or did i miss the right way to preconfigure a filter or its from field)?

**Comment by greg0ire** [comment]
Hi, we try to keep Github issues for bug reports and feature requests only. If you have a question, please ask it on Stack Overflow.

**Comment by RomainSanchez** [comment]
This is an actual feature request as predefined options in the field used in the filter are not taken into account (to the best of my knowledge)

**Comment by greg0ire** [comment]
Ok. Please edit your message to add some punctuation and github flavored markdown… I tried to read it and was like : ![wat](https://i.imgur.com/3Gmiifk.gif)

**Comment by RomainSanchez** [comment]
Ok will do.Nice reference BTW

**Comment by greg0ire** [comment]
Way better, let me re-read that.

**Comment by greg0ire** [comment]
There is a `getFieldOptions()`, is that what you are looking for?

**Comment by RomainSanchez** [comment]
i don't think so.Take a look at `HelperController::retrieveAutocompleteItemsAction()` the callback is retieved from the filter description before any of the filter type methods get called.At this point the only options are the ones passed to `datagridMapper->add()` so the callback that i provided in the form type is ignored.

To make this as simple as possible and avoid blurring the problem with my possibly inaccurate analysis, i made a form type inherited from `ModelAutocompleteType` and provided a custom callback as default option.
If i use this type in a form it works as expected.
When i try to use this same type as field type of a filter the provided callback never gets called.

**Comment by greg0ire** [comment]
[Link for the lazy](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Controller/HelperController.php#L317)

**Comment by greg0ire** [comment]
In the Symfony profiler, in the template panel, there should be a tree showing all the templates involved in displaying your filter. Can you show us the part that you think is relevant to the problem at hand? Also, maybe show us some code.

**Comment by RomainSanchez** [comment]
### The form type:

`
class SearchIndexAutocompleteType extends ModelAutocompleteType
{   
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->setAttribute('callback', $options['callback']);
        $builder->setAttribute('minimum_input_length', $options['minimum_input_length']);
        $builder->setAttribute('items_per_page', $options['items_per_page']);
        $builder->setAttribute('req_param_name_page_number', $options['req_param_name_page_number']);
        $builder->setAttribute(
            'disabled',
            $options['disabled']
            // NEXT_MAJOR: Remove this when bumping Symfony constraint to 2.8+
            || (array_key_exists('read_only', $options) && $options['read_only'])
        );
        $builder->setAttribute('to_string_callback', $options['to_string_callback']);

        if ($options['multiple']) {
            $resizeListener = new ResizeFormListener(
                'hidden', array(), true, true, true
            );

            $builder->addEventSubscriber($resizeListener);
        }
    }

    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $compound = function (Options $options) {
            return $options['multiple'];
        };
        
        $callback = function($admin, $property, $value){
            
            $searchIndex = $admin->getClass() . 'SearchIndex';
            $datagrid = $admin->getDatagrid();
            $queryBuilder = $datagrid->getQuery();
            $alias = $queryBuilder->getRootalias();

            $queryBuilder
                ->leftJoin($searchIndex, 's', 'WITH', $alias . '.id = s.object')
                ->where('s.keyword LIKE :value')
                ->setParameter('value', "%$value%")
            ;
            
           // $datagrid->setValue($property, null, $value);
        };

        $resolver->setDefaults(array(
            'attr' => array(),
            'compound' => $compound,
            'model_manager' => null,
            'class' => null,
            'admin_code' => 'librinfo_email.admin.email',
            'callback' => $callback,
            'multiple' => false,
            'width' => '',
            'context' => '',
            'property' => '',
            'placeholder' => '',
            'minimum_input_length' => 3, //minimum 3 chars should be typed to load ajax data
            'items_per_page' => 10, //number of items per page
            'quiet_millis' => 100,
            'cache' => false,

            'to_string_callback' => null,

            // ajax parameters
            'url' => '',
            'route' => array('name' => 'sonata_admin_retrieve_autocomplete_items', 'parameters' => array()),
            'req_params' => array(),
            'req_param_name_search' => 'q',
            'req_param_name_page_number' => '_page',
            'req_param_name_items_per_page' => '_per_page',

            // CSS classes
            'container_css_class' => '',
            'dropdown_css_class' => '',
            'dropdown_item_css_class' => '',

            'dropdown_auto_width' => false,

            'template' => 'SonataAdminBundle:Form/Type:sonata_type_model_autocomplete.html.twig',
        ));

        $resolver->setRequired(array('property'));
    }

    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return 'blast_search_index_autocomplete';
    }
`

**Comment by greg0ire** [comment]
Dude, you need to learn how Github Flavored Markdown works… I'm not reading that.

**Comment by RomainSanchez** [comment]
Of course didn't mean to post that i was just starting to edit it.

**Comment by RomainSanchez** [comment]
### The filter type

```
class SearchIndexAutocompleteFilter extends ModelAutocompleteFilter
{
    /**
     * {@inheritdoc}
     */
    public function getDefaultOptions()
    {
        $callback = function($admin, $property, $value){
            
            $searchIndex = $admin->getClass() . 'SearchIndex';
            $datagrid = $admin->getDatagrid();
            $queryBuilder = $datagrid->getQuery();
            $alias = $queryBuilder->getRootalias();

            $queryBuilder
                ->leftJoin($searchIndex, 's', 'WITH', $alias . '.id = s.object')
                ->where('s.keyword LIKE :value')
                ->setParameter('value', "%$value%")
            ;
            
           // $datagrid->setValue($property, null, $value);
        };
        
        return array(
            'field_name' => false,
            'field_type' => method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')
                ? 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
                : 'sonata_type_model_autocomplete', // NEXT_MAJOR: Remove ternary (when requirement of Symfony is >= 2.8)
            'field_options' => ['property' => '', 'context' => 'filter', 'callback' => $callback],
            'operator_type' => method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')
                ? 'Sonata\CoreBundle\Form\Type\EqualType'
                : 'sonata_type_equal', // NEXT_MAJOR: Remove ternary (when requirement of Symfony is >= 2.8)
            'operator_options' => array(),
        );
    }
    
    /**
     * {@inheritdoc}
     */
    public function getFieldOptions()
    {
        $callback = function($admin, $property, $value){
            
            $searchIndex = $admin->getClass() . 'SearchIndex';
            $datagrid = $admin->getDatagrid();
            $queryBuilder = $datagrid->getQuery();
            $alias = $queryBuilder->getRootalias();

            $queryBuilder
                ->leftJoin($searchIndex, 's', 'WITH', $alias . '.id = s.object')
                ->where('s.keyword LIKE :value')
                ->setParameter('value', "%$value%")
            ;
            
           // $datagrid->setValue($property, null, $value);
        };
        
        return [
            'property' => '', 
            'context' => 'filter',
            'callback' => $callback
        ];
    }

    /**
     * {@inheritdoc}
     */
    public function getRenderSettings()
    {
        // NEXT_MAJOR: Remove this line when drop Symfony <2.8 support
        $type = method_exists('Symfony\Component\Form\AbstractType', 'getBlockPrefix')
            ? 'Sonata\AdminBundle\Form\Type\Filter\DefaultType'
            : 'sonata_type_filter_default';

        return array($type, array(
            'field_type' => 'blast_search_index_autocomplete',
            'field_options' => $this->getFieldOptions(),
            'operator_type' => $this->getOption('operator_type'),
            'operator_options' => $this->getOption('operator_options'),
            'label' => $this->getLabel(),
        ));
    }

}
```

As you can see i tried to pass the callback in `getdefaultOptions()` and `getFieldOptions()`

### Filter type service

```
blast_base_entities.filter.type.search_index_autocomplete_filter:
        class: Blast\BaseEntitiesBundle\Filter\SearchIndexAutocompleteFilter
        tags:
            - { name: sonata.admin.filter.type, alias: blast_search_index_autocomplete_filter }
```

**Comment by greg0ire** [comment]
Are you aware you can edit your messages?

**Comment by greg0ire** [comment]
> In the Symfony profiler, in the template panel, there should be a tree showing all the templates involved in displaying your filter. Can you show us the part that you think is relevant to the problem at hand? 

You should really address that. It's hard to help without knowing what templates are used with.

**Comment by RomainSanchez** [comment]
here is the rendering call graph for the filter

```
SonataAdminBundle:CRUD:base_list.html.twig::block(list_filters) 15.67ms/7%
│             │ └ form_div_layout.html.twig::block(form_errors)
│             │ └ SonataAdminBundle:CRUD:base_list.html.twig::block(sonata_list_filter_group_class)
│             │ └ form_div_layout.html.twig::block(choice_widget)
│             │ │ └ SonataAdminBundle:Form:filter_admin_fields.html.twig::block(choice_widget_collapsed)
│             │ │   └ form_div_layout.html.twig::block(choice_widget_collapsed)
│             │ │     └ form_div_layout.html.twig::block(widget_attributes)
│             │ │     └ form_div_layout.html.twig::block(choice_widget_options)
│             │ └ BlastBaseEntitiesBundle:Form:fields.html.twig::block(blast_search_index_autocomplete_widget) 1.37ms/1%
│             │ │ └ SonataAdminBundle:Form:filter_admin_fields.html.twig::block(sonata_type_model_autocomplete_widget) 1.36ms/1%
│             │ │   └ SonataAdminBundle:Form/Type:sonata_type_model_autocomplete.html.twig
│             │ │     └ SonataAdminBundle:Form/Type:sonata_type_model_autocomplete.html.twig::block(sonata_type_model_autocomplete_ajax_request_parameters)
│             │ │     └ SonataAdminBundle:Form/Type:sonata_type_model_autocomplete.html.twig::block(sonata_type_model_autocomplete_dropdown_item_format)
│             │ │     └ SonataAdminBundle:Form/Type:sonata_type_model_autocomplete.html.twig::block(sonata_type_model_autocomplete_
```

**Comment by RomainSanchez** [comment]
> ok it's better. Still improvable though, I added syntactic coloration myself, have a look.

Well noted

**Comment by RomainSanchez** [comment]
Hello?

**Comment by greg0ire** [comment]
If I were you I'd have a look at `SonataAdminBundle:Form:filter_admin_fields.html.twig` and see if it looks like options can be passed to filters here.
